File: 设计模式/程序/20-ObserverPattern/ObserverPattern(C++)/src/Observer.cpp
#include "Observer.h"
#include <bitset>
#include <iostream>


namespace ObserverPatternApp
{
    void BinaryObserver::Update(int newState) const
    {
        std::cout << "Binary String: " << (std::bitset<4>)newState << std::endl; 
    }

    void OctalObserver::Update(int newState) const
    {
        std::cout << "Octal String: " << std::oct << newState << std::endl; 
    }

    void HexaObserver::Update(int newState) const
    {
        std::cout << "Hexa String: " << std::hex << newState << std::endl; 
    }
}

File: 设计模式/程序/20-ObserverPattern/ObserverPattern(python)/Observer.py
import abc

# 观察者基类
class Observer(metaclass = abc.ABCMeta):
    @abc.abstractmethod
    def Uptate(self, newState):
        pass

# 观察者实现
class BinaryObserver(Observer):
    def __init__(self):
        pass
    def Uptate(self, newState):
        print('Binary String: {0}'.format(bin(newState)))

class OctalObserver(Observer):
    def __init__(self):
        pass
    def Uptate(self, newState):
        print('Octal String: {0}'.format(oct(newState)))

class HexaObserver(Observer):
    def __init__(self):
        pass
    def Uptate(self, newState):
        print('Hexa String: {0}'.format(hex(newState)))